streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/ImplicitConversions.scala

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.scala
+
+import org.apache.kafka.streams.kstream.{KStream => KStreamJ,
+  KTable => KTableJ,
+  KGroupedStream => KGroupedStreamJ,
+  SessionWindowedKStream => SessionWindowedKStreamJ,
+  TimeWindowedKStream => TimeWindowedKStreamJ,
+  KGroupedTable => KGroupedTableJ, _}
+
+import org.apache.kafka.streams.scala.kstream._
+import org.apache.kafka.streams.{ KeyValue, Consumed }
+import org.apache.kafka.common.serialization.Serde
+
+import scala.language.implicitConversions
+
+/**
+ * Implicit conversions between the Scala wrapper objects and the underlying Java
+ * objects.
+ */
+object ImplicitConversions {
+
+  implicit def wrapKStream[K, V](inner: KStreamJ[K, V]): KStream[K, V] =
+    new KStream[K, V](inner)
+
+  implicit def wrapKGroupedStream[K, V](inner: KGroupedStreamJ[K, V]): KGroupedStream[K, V] =
+    new KGroupedStream[K, V](inner)
+
+  implicit def wrapSessionWindowedKStream[K, V](inner: SessionWindowedKStreamJ[K, V]): SessionWindowedKStream[K, V] =
+    new SessionWindowedKStream[K, V](inner)
+
+  implicit def wrapTimeWindowedKStream[K, V](inner: TimeWindowedKStreamJ[K, V]): TimeWindowedKStream[K, V] =
+    new TimeWindowedKStream[K, V](inner)
+
+  implicit def wrapKTable[K, V](inner: KTableJ[K, V]): KTable[K, V] =
+    new KTable[K, V](inner)
+
+  implicit def wrapKGroupedTable[K, V](inner: KGroupedTableJ[K, V]): KGroupedTable[K, V] =
+    new KGroupedTable[K, V](inner)
+
+  implicit def tuple2ToKeyValue[K, V](tuple: (K, V)): KeyValue[K, V] = new KeyValue(tuple._1, tuple._2)
```

Comment:
@debasishg Ping on this comment again, could you elaborate if my concern is valid or not?

Reply:
@guozhangwang I am not sure I understand your concern. The only purpose of this implicit is to allow an implicit conversion from `Tuple2` to `KeyValue(key, value)`. Just a helper which we found useful in many cases for developing applications or tests.

Reply:
Here is my concern: in `map` and `flatMap`, we call

```
mapper.tupled andThen tuple2ToKeyValue
```

Does that mean that for each pair of K, V pair parameters, we would first construct a `Tuple2` object of this case class, and then  apply the mapper, and then create a new `KeyValue` from the result `Tuple2` object? If that is true, then we are creating a short-lived object for each record processed in `map`. I'm not sure if it will have a pressure on the GC.

Reply:
@guozhangwang - You are correct that with the current implementation there will be `Tuple2`s created. But it's difficult to say if there will be GC pressure. For that we need to analyze runtime behaviors and see what the JIT does. There's of course a way we can fall back to the implementation which does less allocation ..

```
def map[KR, VR](mapper: (K, V) => (KR, VR)): KStream[KR, VR] = {
  val mapperJ: KeyValueMapper[K, V, KeyValue[KR, VR]] = ((k: K, v: V) => {
    val res = mapper(k, v)
    new KeyValue[KR, VR](res._1, res._2)
  }).asKeyValueMapper
  inner.map[KR, VR](mapperJ)
}
```

We did run some tests in bulk to check the diff in performance between the 2 versions. Couldn't find much of a difference though.